feat(posts): a single benchmark number is a lie (series 2/5) (#313)
kakkoyun/me: content/posts/go-benchmarks-lying-statistics.md

---
title: "A Single Benchmark Number Is a Lie"
description: "One benchmark run is one sample from a distribution you haven't seen. Go's benchstat and a twenty-line awk script give you the tools to ask whether your results are real — and whether your environment is stable enough to answer."
date: 2026-09-08T00:00:00Z

> publishDate: 2026-09-08T00:00:00Z

Author: New slot: **Friday, August 28** (moved from Tue Sep 8), part 2 of the series' Friday run.

```suggestion
date: 2026-08-28T00:00:00Z
publishDate: 2026-08-28T00:00:00Z
```

---
_Generated by [Claude Code](https://claude.ai/code)_

> publishDate: 2026-09-08T00:00:00Z

Author: Supersedes my earlier Aug 28 suggestion (not applied, so this replaces it). **Fri Sep 4**, 2/5.

```suggestion
date: 2026-09-04T00:00:00Z
publishDate: 2026-09-04T00:00:00Z
```

---
_Generated by [Claude Code](https://claude.ai/code)_

#### Comparing Two Runs

Single-file summaries show how consistent a run set is. The comparison mode is more useful:

```bash
go test -bench=. -benchmem -count=20 -benchtime=1s . > old.txt
# make your change
go test -bench=. -benchmem -count=20 -benchtime=1s . > new.txt
benchstat old.txt new.txt
```

Here is the comparison between the idle run and the loaded run from the demo results — `sec/op` only, allocations were identical:

```text
                      │ results/idle.txt │           results/noisy.txt           │
                      │      sec/op      │    sec/op     vs base                 │
MakeBuffer_Correct-16        11.32n ± 5%   37.40n ± 25%  +230.34% (p=0.000 n=20)
```

Three numbers: the delta (+230.34%, meaning the loaded machine ran the same benchmark three times slower), the p-value (p=0.000, effectively zero; the difference is real), and the sample count (n=20 per side).

When the p-value exceeds 0.05, benchstat prints `~` instead of a delta — the correct result, meaning "no measurable difference with this sample size." Report it as such rather than rerunning.

---

### What Benchstat Doesn't Tell You

> The `± 5%` and `± 25%` figures hint at spread, but they describe how tightly the geomean is estimated, not whether your *environment* is stable enough to trust any comparison you run in it. For that you need a different number: the **coefficient of variation**.

Author: This contradicts your own (correct) sidenote 40 lines up. The sidenote says per-benchmark rows show the **median** and that geomean appears only in the cross-benchmark summary row — right. Then this sentence claims the `±%` describes "how tightly the geomean is estimated" — wrong on both counts: it's not about the geomean, and it's a spread indicator (range of the non-outlier samples around the median), not a confidence interval. Line 145 repeats the geomean slip ("estimate the geomean with confidence").

```suggestion
The `± 5%` and `± 25%` figures hint at spread — the range of the sampled runs around the median — but they do not tell you whether your *environment* is stable enough to trust any comparison you run in it. For that you need a different number: the **coefficient of variation**.
```

---
_Generated by [Claude Code](https://claude.ai/code)_

#### The Shape Behind the Mean

The arithmetic mean of 34.97 ns/op in the noisy condition looks unremarkable until you inspect the raw runs. The first seven land between 38 and 44 ns/op. Runs eight through sixteen drop to 25 to 29 ns/op, almost half the speed of the first cluster. Then the last four drift back toward 38 to 39. The mean of 34.97 sits between two populations that never actually existed as stable states.

A single number — any single number — hides this completely. This is why the FOSDEM post's advice on strip plots applies here: one dot per measurement reveals the bimodality immediately. Boxplots would obscure it; a mean buries it entirely.

The CV of 18.88% catches the problem numerically. But the raw data is showing you something worse than noise: the measurement environment was shifting between two regimes during the run. That is not recoverable with more statistics. The third condition in the table — a pinned container — is the start of the answer. Post 3 ([Local Reproduction](/posts/go-benchmarks-lying-local-reproduction/)) covers what isolation actually buys you.

---

### How Many Runs Are Enough?

> `-count=10` is the practical floor. Benchstat needs enough samples to estimate the geomean with confidence; below ten the estimates become unreliable. Use `-count=20` for anything you intend to report or check into a CI baseline. [Kalibera and Jones (ECOOP 2013)](https://dl.acm.org/doi/10.1145/2509136.2509184) establish that N ≥ 30 is needed for robust inter-run statistics; 20 is a reasonable engineering compromise between rigor and machine time.

Author: Citation is scrambled three ways here (and repeated in Resources, line 221):

1. **Venue**: Kalibera & Jones's *Rigorous Benchmarking in Reasonable Time* is **ISMM 2013** (International Symposium on Memory Management), not ECOOP.
2. **DOI**: `10.1145/2509136.2509184` is in the OOPSLA/SPLASH 2013 proceedings block — it points at some other paper. The ISMM '13 paper's DOI is `10.1145/2464157.2464160`.
3. **Attribution strength**: "establish that N ≥ 30 is needed" over-claims — the paper's contribution is a *method* for choosing repetition counts from measured variance, not a universal 30-run floor. Suggest "argue for statistically principled repetition counts" and keep your "20 is an engineering compromise" framing, which is honest.

Also fix the same venue/DOI in the Resources entry. Also: "Benchstat needs enough samples to estimate the geomean with confidence" in this line — same geomean slip as line 99; it's the median in per-benchmark rows.

---
_Generated by [Claude Code](https://claude.ai/code)_

This is p-hacking, and it invalidates everything benchstat reports. The p-value is calibrated for a pre-specified number of experiments, not an open-ended search. With `-count=20` and α=0.05, you expect approximately one false positive for every twenty benchmarks by chance alone. Select for runs that clear the threshold and you have systematically chosen flukes.

The discipline is simple but requires commitment up front:
1. Decide on N before running.
2. Run once.
3. Report what benchstat says, including the `~` results.

If the result is not significant with your chosen N, you have two honest options: accept "no measurable difference" or pre-commit to a larger N and run exactly once more. What you cannot do is rerun after seeing a disappointing p-value.

Recognizing it in practice:

- "I ran it a few times until it stabilised": p-hacking.
- "The improvement only showed up on the third attempt": p-hacking.
- "benchstat showed improvement when I closed my editor": environmental confounding, equally problematic.

---

### Effect Size vs Statistical Significance

A low p-value tells you the difference is unlikely to be zero. It does not tell you the difference matters.

> With 100 samples, a 0.3% change will clear p=0.05 with ease. A 0.3% improvement to an HTTP handler running at 200 µs saves 600 picoseconds per request. Statistically real, practically irrelevant. The code complexity required to achieve it will cost more in maintenance than it saves in latency.

Author: **Math error, 1000× off**: 0.3% of 200 µs = 0.6 µs = **600 nanoseconds**, not picoseconds. In a post whose whole thesis is numerical honesty, this one line would get quoted back at you.

```suggestion
With 100 samples, a 0.3% change will clear p=0.05 with ease. A 0.3% improvement to an HTTP handler running at 200 µs saves 600 nanoseconds per request. Statistically real, practically irrelevant. The code complexity required to achieve it will cost more in maintenance than it saves in latency.
```

---
_Generated by [Claude Code](https://claude.ai/code)_
